### ISSUE TITLE
Null Object Pattern

### ISSUE THREAD

**Issue by thekensman** [pydantic/pydantic#979]
# Feature Request

Please complete:
* OS: **Ubuntu 19.04**
* Python version `import sys; print(sys.version)`: **3.7.3**
* Pydantic version `import pydantic; print(pydantic.VERSION)`: **1.0b2**

Hi,

I've been using pydantic for several months now and love it. One thing I'm trying to implement in my APIs though is the [Null Object Pattern](https://sourcemaking.com/design_patterns/null_object). This basically means I'd like to always return the same interface, even on blank objects. Ex.
```
{"example": {"blank_data": None}}
```
Instead of:
```
{"example": None}
```
This allows APIs to reduce null checking and returns less variation in responses. This seems simple (create a default version of of the field for the default, like `default=MyField()`). However, it gets very complicated if your field contains required values.

I've created a function to return null objects for these default values:

```py
from pydantic import BaseModel, ValidationError

def get_null_object(field):
    def nullify_field_type(field):
        #field if this is called while declaring a model
        #field.type_ if this is called while instantiating a model and after
        class null_field_type(field.type_ if getattr(field, "type_", None) else field): pass

        for sub_field_name, sub_field in null_field_type.__fields__.items():
            #Recurse on sub models
            if issubclass(sub_field.type_, PydanticBaseModel):
                sub_field.type_ = nullify_field_type(sub_field)

            sub_field.default = None
            sub_field.field_info.default = None
            sub_field.required = False

        return null_field_type

    return nullify_field_type(field)()

class Base(BaseModel):
    optional_type: str = "optional"
    required_type: str = ...

class NestedBase(BaseModel):
    base: Base = get_null_object(Base)

class TestModel(BaseModel):
    base: Base = get_null_object(Base)
    nested_base: NestedBase = get_null_object(NestedBase)

print(TestModel()) #TestModel base=<null_field_type optional_type=None required_type=None> nested_base=<null_field_type base=None>
print(get_null_object(TestModel.__fields__["base"])) #null_field_type optional_type=None required_type=None
print(get_null_object(TestModel.__fields__["nested_base"])) #null_field_type base=None
try:
    TestModel(base={"optional_type": "no required type!"})
except ValidationError as ve:
    print("doesn't work")

try:
    TestModel(nested_base={"base": {"optional_tyoe": "no required type!"}})
except ValidationError as ve:
    print("doesn't work")
```
I wonder if it would make sense to add this to pydantic as a helper method? 
From there 
- We could specify that the field should be a null object through the config, removing the need to re-specify the type of the null object.
- We could add support for converting None values into null objects if specified (ex. `TestModel(base=None, nested_base=None)`)

Anyway not sure if you'd like me to PR this but I'll keep developing it for my personal use!

**Comment by dmontagu** [comment]
I think there's something interesting here, but it seems like it might be awkward for practical use with OpenAPI / JSON schema.

Before it makes sense to integrate this into pydantic, I think at least two of the following three things should be the case:
1. There is a lot of interest in the feature
2. The feature has good compatibility with OpenAPI/JSON schema
3. The feature has a clean, simple, easily-maintained implementation

------

I think an approach that might help push this in an easier-to-reason-about direction would be to create a generic class like `NullObject[T]`, and implement it so that you can get this null-object pattern by replacing `Optional[T]` with `NullObject[T]` where appropriate in your model declarations.

So it would look like:
```python
class TestModel(BaseModel):
    base: NullObject[Base]
    nested_base: NullObject[NestedBase]
```

This could be implemented in a *similar* way to `GenericModel` (though I don't think you could use `GenericModel` directly). In particular, you'd want a way to generate a model for the value when the object is null using `create_model`, and then `NullObject.__class_getitem__` should return `Union[T, create_null_model(T)]`.

Let me know if you think this sounds interesting and want more direction along these lines.

**Comment by thekensman** [comment]
Oh wow I think that's a great idea. That lets you build the typing right into the declaration. Also a good point about the OpenAPI schema being affected.

Let me try to put something together along those lines. For now it looks like I can build this on top of pydantic without messing with the internals, which is good. Let me see what I can come up with - if no one else is interested it'll at least make my own APIs much cleaner.

Definitely appreciate the direction!

**Comment by thekensman** [comment]
I got pretty far on an implementation here:
https://github.com/thekensman/pydantic/blob/null_object/tests/test_null_object.py
https://github.com/thekensman/pydantic/blob/null_object/pydantic/null_object.py

Before I started doing some soul searching on the null object pattern. It's fit a lot of my needs in the past but thinking deeper into it this pattern can make some api responses confusing to parse. Ex. 
You have a list of classes that each have an optional teacher object. Traditionally this could look like:
```[{"name": "Study Hall", "teacher": None}, {"name": "Trigonometry", "teacher": {"id": 1, "name": "Teacher", "age": 40, "accredited": True}}, {"id": 2, "name": "Geography", "teacher": {"name": "Teacher2", "age": null, "accredited": False}}]```
With the null object pattern this would look like:
```[{"name": "Study Hall", "teacher": {"id": null, "name": null, "age": null, "accredited", null}}, {"name": "Trigonometry", "teacher": {"name": "Teacher", "age": 40, "accredited": True}}, {"name": "Geography", "teacher": {"name": "Teacher2", "age": null, "accredited": False}}]```
Now if you wanted to send a note to all classes that have a teacher who didn't fill out their age to remind them they need to do it... Suddenly study hall is getting a note unless the client KNOWS to do a null check on the object, which lets you shoot yourself in the foot if you forget.
This gets more obvious if you expect a teacher to always have an id, but the null object has an id None.

So even though it's going to lead to things like:
```teacher = class["teacher"]
if teacher and not teacher["age"]:...```
It seems like this is information you'd like to know without parsing a potentially empty object

Anyway I decided to put my APIs back to the more traditional `"object": None` for now... Just wish I'd thought this through before getting most of the way through the implementation!
